cfdp-core/src/segments.rs

```diff
@@ -0,0 +1,211 @@
+use std::cmp::Ordering;
+
+// update the contiguous segment list with the new segment, filling any gap (i.e. unifying segments) if required
+pub fn update_segments(v: &mut Vec<(u64, u64)>, seg: (u64, u64)) {
+    if seg.0 >= seg.1 {
+        panic!("Invalid segment {:?}", seg);
+    }
```

Comment:
Instead of a `panic` here, would you mind changing the signature to return an Result and error if start is beyond the end for the new segment?  I don't think as implemented it will happen to the engine but at least this way it could be propagated up to a log message even if the program chooses exit at that case.

This would probably require a new Error type in this module to handle that.

Reply:
If I understand anything from the Rust panic vs result is that when the caller makes a mistake it's ok to panic.
from the book https://doc.rust-lang.org/book/ch09-03-to-panic-or-not-to-panic.html:

> Panicking when the contract is violated makes sense because a contract violation always indicates a caller-side bug and it’s not a kind of error you want the calling code to have to explicitly handle. In fact, there’s no reasonable way for calling code to recover; the calling programmers need to fix the code.

In our case - the caller calls the update with an invalid segment - that is  a segment having the end before the start. There is no other way to handle this that fixing the caller code.
I could replace the "if panic" with std::assert, maybe that is better.

Reply:
Ah I see, definitely understand the logic, I think panic just scares me :laughing: Maybe you can add a `#[should_panic]` test then?

This will panic also for a FileDataPDU with 0 length, which if I understand the standard shouldn't be constructed/sent to begin with and would fall into the "the other engine did something bad" list right?

Reply:
That's a good catch - we have to log and reject PDUs with length 0. That is definitely not the same thing because if a remote entity sends those we should not panic - they may be buggy but that is not our concern.

I will add a check for that and also a should_panic test. 

I noticed several PDUs which make the cfdp daemon quit - for example PDUs containing unknown entity ids. We should also just log and reject those instead of panicking or quitting.

Reply:
I agree that individual PDUs should not make the daemon quit. Probably oversight on my side and not having done as much real word testing as you have. I'll make an issue to help track that.

Reply:
> Ah I see, definitely understand the logic, I think panic just scares me laughing Maybe you can add a `#[should_panic]` test then?
> 
> This will panic also for a FileDataPDU with 0 length, which if I understand the standard shouldn't be constructed/sent to begin with and would fall into the "the other engine did something bad" list right?

thinking more about it I will replace "if..panic" with an assert. The rationale is that the segment update method is not part of public API but internal thing - so really it's just checking an invariant for safety.

Reply:
As long as we're guarding against the assertion upstream I think that should be a fine approach. Guarding against 0 length PDUs should ensure we never trigger that mode.